### PR TITLE
Remove IServerLoader and add the IServerFactory to the DI container

### PR DESCRIPTION
**Pull request by davidfowl**
- This change removes the indirection and between the IServerLoader and
the IServerFactory. We now add the IServerFactory directly to the DI
container and resolve it from there.

/cc @lodejard @Tratcher @JunTaoLuo

### CODE REVIEW

File: src/Microsoft.AspNetCore.Hosting/WebHostBuilderExtensions.cs [hunk 1]
```diff
@@ -101,7 +104,46 @@ public static IWebHostBuilder UseServer(this IWebHostBuilder hostBuilder, string
...
+                .FirstOrDefault();
+
+            if (serverTypeInfo == null)
+            {
+                throw new InvalidOperationException($"No server type found that implements IServerFactory in assembly: {assemblyName}.");
+            }
+
+            return hostBuilder.ConfigureServices(services =>
+            {
+                services.AddSingleton(typeof(IServerFactory), serverTypeInfo.AsType());
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Hosting/WebHostBuilderExtensions.cs]: Transient?

Comment (Tratcher) [inline src/Microsoft.AspNetCore.Hosting/WebHostBuilderExtensions.cs]: `UseServer<TServerFactory>() where TServerFactory is IServerFactory`?

File: src/Microsoft.AspNetCore.Hosting/WebHostBuilder.cs [hunk 2]
```diff
@@ -248,6 +222,12 @@ private IServiceCollection BuildHostingServices()
             services.AddSingleton(defaultPlatformServices.Application);
             services.AddSingleton(defaultPlatformServices.Runtime);
 
+            if (!string.IsNullOrEmpty(_options.ServerFactoryAssembly))
+            {
+                // Add the server factory
+                this.UseServer(_options.ServerFactoryAssembly);
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Hosting/WebHostBuilder.cs]: It's a little strange to go back and call one of the extensions here. Revert the changes to `UseServer(string)` and just have a private `ResolveServerAssembly(string, IServiceCollection)`.

File: test/Microsoft.AspNetCore.Hosting.Tests/WebHostBuilderTests.cs [hunk 3]
```diff
@@ -237,99 +237,7 @@ public void ConfigureServices_CanBeCalledMultipleTimes()
             Assert.NotNull(host.Services.GetRequiredService<ServiceA>());
             Assert.NotNull(host.Services.GetRequiredService<ServiceB>());
         }
-
-        [Fact]
-        public void CodeBasedSettingsCodeBasedOverride()
-        {
-            var hostBuilder = new WebHostBuilder()
```
Comment (Tratcher) [inline test/Microsoft.AspNetCore.Hosting.Tests/WebHostBuilderTests.cs]: These were real config tests (not server tests). How about converting it to test a different field like CaptureStartupErrors?

File: src/Microsoft.AspNetCore.Hosting/WebHostBuilderExtensions.cs (hunk 1 above)
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Hosting/WebHostBuilderExtensions.cs]: Can't be transient. See the other place where we register this. Unless we remove the instance overloads it's bad to use different lifetimes for the same interface.

File: test/Microsoft.AspNetCore.Hosting.Tests/WebHostBuilderTests.cs (hunk 3 above)
Comment (davidfowl) [inline test/Microsoft.AspNetCore.Hosting.Tests/WebHostBuilderTests.cs]: Hmm, will check that.

**Comment by davidfowl** [comment]
:up::date:

File: src/Microsoft.AspNetCore.Hosting/WebHostBuilderExtensions.cs [hunk 4]
```diff
@@ -101,7 +104,33 @@ public static IWebHostBuilder UseServer(this IWebHostBuilder hostBuilder, string
                 throw new ArgumentNullException(nameof(assemblyName));
             }
 
-            return hostBuilder.UseSetting(WebHostDefaults.ServerKey, assemblyName);
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Hosting/WebHostBuilderExtensions.cs]: Not just going to revert this change and let it happen in Build?

**Comment by Tratcher** [comment]
:shipit:

File: src/Microsoft.AspNetCore.Hosting/WebHostBuilderExtensions.cs (hunk 4 above)
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Hosting/WebHostBuilderExtensions.cs]: Ah, will do that.
